### PR TITLE
[Bug] (fix) DeltaWriter::mem_consumption() coredump

### PR DESCRIPTION
**Pull request by chenlinzhong**
# Proposed changes

Issue Number: open  #9237

## Problem Summary:

- mem_consumption() returned value may be changed during sort period for same element, this does not meet comp requirements
- make snapshot for  memory consumption  before call std::sort and used snapshot where start call std::sort

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
3. Has document been added or modified: (Yes/No/No Need)
4. Does it need to update dependencies: (Yes/No)
5. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/runtime/tablets_channel.cpp [hunk 1]
```diff
@@ -216,12 +216,12 @@ Status TabletsChannel::reduce_mem_usage(int64_t mem_limit) {
     // Sort the DeltaWriters by mem consumption in descend order.
     std::vector<DeltaWriter*> writers;
     for (auto& it : _tablet_writers) {
+        it.second->save_mem_consumption_snapshot();
```
Comment (yiguolei) [inline be/src/runtime/tablets_channel.cpp:219]: What's the core reason?

Comment (chenlinzhong) [reply be/src/runtime/tablets_channel.cpp:219]: > What's the core reason?

mem_consumption() returned value may be changed during sort period for same element, this does not meet comp requirements

**Comment by HappenLee** [comment]
why the case normal in old code?

**Review by HappenLee — DISMISSED** [review]
LGTM

**Review by yiguolei — DISMISSED** [review]
LGTM

**Review by yiguolei — DISMISSED** [review]
LGTM

**Review by morningman — APPROVED** [review]

**Pull request merged**